pallets/phala/src/stakepool.rs

```diff
@@ -878,17 +935,27 @@ pub mod pallet {
 			let miner: T::AccountId = pool_sub_account(pid, &worker);
 			// Mining::stop_mining will notify us how much it will release by `on_stopped`
 			<mining::pallet::Pallet<T>>::stop_mining(miner)?;
+			ensure!(
```

Comment:
Be careful! The very first rule of pallet development is to make sure to not modify the state if the return value is `Err(..)`. Usually this is achieved by doing all the checks once together ever before starting to write to the storage. It's elegant and efficient.

Here if `stop_mining` runs successfully, it has already changed the states in the Mining pallet. However this `ensure!` may return an error, resulting the state modified.

Unit test also provides another layer of protection. You should cover all the `Err` case with `assert_noop!()` check. It checks the error message, as well as the storage is unmodified.

In some very complex case, it's ok to add a `#[transactional]` attribute to the function so that it will automatically rollback the changes if it returns an `Err`.